iib/workers/tasks/opm_operations.py

```diff
@@ -0,0 +1,169 @@
+            log.info('Port %d is in use, trying another.', port)
+            port += 1
+        sock.close()
+
+    err_msg = f'No free port has been found after {port_end - port_start} attempts.'
+    log.error(err_msg)
+    raise IIBError(err_msg)
+
+
+def opm_serve_from_index(base_dir, from_index):
+    """
+    Locally start OPM registry service, which can be communicated with using gRPC queries.
+
+    Due to IIB's paralellism, the service can run multiple times, which could lead to port
+    binding conflicts. Resolution of port conflicts is handled in this function as well.
+
+    :param str base_dir: base directory to create temporary files in.
+    :param str from_index: index image to inspect.
+    :return: tuple containing port number of the running service and the running Popen object.
+    :rtype: (int, Popen)
+    :raises IIBError: if all tried ports are in use, or the command failed for another reason.
+    """
+    from iib.workers.tasks.build import _get_index_database
+
+    if not is_image_fbc(from_index):
+        db_path = _get_index_database(from_index, base_dir)
+        return opm_registry_serve(db_path)
+
+    catalog_dir = get_catalog_dir(from_index, base_dir)
+    return opm_serve(catalog_dir)
+
+
+def opm_serve(catalog_dir):
+    """
+    Locally start OPM service, which can be communicated with using gRPC queries.
+
+    Due to IIB's paralellism, the service can run multiple times, which could lead to port
+    binding conflicts. Resolution of port conflicts is handled in this function as well.
+
+    :param str catalog_dir: path to file-based catalog directory that should be served.
+    :return: tuple containing port number of the running service and the running Popen object.
+    :rtype: (int, Popen)
+    :raises IIBError: if all tried ports are in use, or the command failed for another reason.
+    """
+    conf = get_worker_config()
+    port_start = conf['iib_grpc_start_port']
+    port_end = port_start + conf['iib_grpc_max_port_tries']
+
+    port = _get_free_port(port_start, port_end)
+    cmd = ['opm', 'serve', catalog_dir, '-p', str(port), '-t', '/dev/null']
+    cwd = os.path.abspath(os.path.join(catalog_dir, os.path.pardir))
+    return (
+        port,
+        _serve_cmd_at_port(
+            cmd, cwd, port, conf['iib_grpc_max_tries'], conf['iib_grpc_init_wait_time']
+        ),
+    )
+
+
+def opm_registry_serve(db_path):
+    """
+    Locally start OPM registry service, which can be communicated with using gRPC queries.
+
+    Due to IIB's paralellism, the service can run multiple times, which could lead to port
+    binding conflicts. Resolution of port conflicts is handled in this function as well.
+
+    :param str db_path: path to index database containing the registry data.
+    :return: tuple containing port number of the running service and the running Popen object.
+    :rtype: (int, Popen)
+    :raises IIBError: if all tried ports are in use, or the command failed for another reason.
```

Comment:
Is this function raises IIBError or _serve_cmd_at_port()

Reply:
Removed, thanks